### DISCUSSION TITLE
Teleported x-data using async alpine broken on Alpine ^v3.14.4

### DISCUSSION THREAD

**Discussion by danie-ramdhani** [alpinejs/alpine#4467] [7. Bugs (archived)] [answered]
Hi i'm using [async alpine](https://github.com/Accudio/async-alpine) and after Alpine.js ^v3.14.4 it broke the teleported `x-data` component using async alpine.

here's the minimal bug reproduction:

index.html (try change to alpine version ^v3.14.4 to v3.14.3 then see the console)

```html
<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="UTF-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <title>Document</title>
        <script defer src="https://cdn.jsdelivr.net/npm/async-alpine@2.x.x/dist/async-alpine.script.js"></script>
        <!-- <script src="https://cdn.jsdelivr.net/npm/alpinejs@3.14.3/dist/cdn.min.js" defer></script> -->
        <!-- <script src="https://cdn.jsdelivr.net/npm/alpinejs@3.14.4/dist/cdn.min.js" defer></script> -->
        <script src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js" defer></script>

        <script src="https://cdn.tailwindcss.com"></script>
    </head>
    <body class="flex items-center justify-center w-full h-dvh">
    
        <div x-data="{ open: false }">
            <button type="button" x-on:click="open = !open">Toggle</button>
            
            <template x-teleport="body">
                <div
                    class="h-dvh fixed inset-0 z-50 w-full bg-slate-950/50 flex items-center justify-center"
                    x-show="open"
                    x-cloak
                    x-on:click="open =! open"
                >
                    <div x-load x-load-src="async-alpine.js" x-ignore x-data="asyncAlpineData" class="bg-pink-500 text-white"></div>
                    <div
                        x-data="{ 
                        init(){
                            console.log('inline alpine data')
                        }
                     }"
                    ></div>
                </div>
            </template>
        </div>
        
    </body>
</html>
```

async-alpine.js

```js
export default function asyncAlpineData() {
    return {
        init() {
            this.$root.innerHTML = "foo";
            console.log("hello");
        },
    };
}
```

**Comment by ekwoka** [comment]
fixed in #4450

**Reply by danie-ramdhani** [reply]
i think it's not.

<img width="700" alt="image" src="https://github.com/user-attachments/assets/73bfebe7-db15-4b78-96b6-50f9d8eb2455">

<img width="611" alt="image" src="https://github.com/user-attachments/assets/e57a7c33-c549-44bc-b1ed-a7f41fdde9f1">

**Reply by danie-ramdhani** [reply]
and i don't know what's going on with my livewire modal when i hit server request, the modal keep hidden because `x-cloak`, so i need to do something like this and it's back to normal:

```js
this.$watch('open', value => {
    value ? this.$root.removeAttribute('x-cloak') : this.$root.setAttribute('x-cloak', '');
});
```

**Reply by ekwoka** [reply]
it's inside a template, it doesn't need x-cloak at all.

**Reply by danie-ramdhani** [reply]
idk, it "blip" when i'm not using `x-cloak` (or maybe that's teleport behavior on livewire ?)

**Reply by ekwoka** [reply]
I don't get any blip...

not in normal Alpine. Since it should clone the element and evaluate the x-show in the same frame...

**Reply by danie-ramdhani** [reply]
[Video_2024_12_04-1.webm](https://github.com/user-attachments/assets/5bca9653-09f8-4837-a4d3-29ed55e126da)

mmmm.... maybe it's alpine teleport behavior on livewire

**Reply by danie-ramdhani** [reply]
well... back to the main problem. async alpine teleported component is still broken in the latest alpine version. and i don't know how to downgrade alpine.js only because it's bundled with livewire.

**Reply by ekwoka** [reply]
It might be something to point out to that project.

I don't know enough about how it works, but as far as I can check...expected Alpine behaviors are working in 3.14.6...

Maybe downgrade Livewire?

**Reply by danie-ramdhani** [reply]
Have you tried the code ? (Try alpine ver 3.14.6 then 3.14.3 to see the different)

```html
<!DOCTYPE html>
<html lang="en">
    <head>
        <meta charset="UTF-8" />
        <meta name="viewport" content="width=device-width, initial-scale=1.0" />
        <title>Document</title>
        <script defer src="https://cdn.jsdelivr.net/npm/async-alpine@2.x.x/dist/async-alpine.script.js"></script>
        <!-- <script src="https://cdn.jsdelivr.net/npm/alpinejs@3.14.3/dist/cdn.min.js" defer></script> -->
        <!-- <script src="https://cdn.jsdelivr.net/npm/alpinejs@3.14.4/dist/cdn.min.js" defer></script> -->
        <script src="https://cdn.jsdelivr.net/npm/alpinejs@3.x.x/dist/cdn.min.js" defer></script>

        <script src="https://cdn.tailwindcss.com"></script>
    </head>
    <body class="flex items-center justify-center w-full h-dvh">
    
        <div x-data="{ open: false }">
            <button type="button" x-on:click="open = !open">Toggle</button>
            
            <template x-teleport="body">
                <div
                    class="h-dvh fixed inset-0 z-50 w-full bg-slate-950/50 flex items-center justify-center"
                    x-show="open"
                    x-cloak
                    x-on:click="open =! open"
                >
                    <div x-load x-load-src="async-alpine.js" x-ignore x-data="asyncAlpineData" class="bg-pink-500 text-white"></div>
                    <div
                        x-data="{ 
                        init(){
                            console.log('inline alpine data')
                        }
                     }"
                    ></div>
                </div>
            </template>
        </div>
        
    </body>
</html>
```

```js
export default function asyncAlpineData() {
    return {
        init() {
            this.$root.innerHTML = "foo";
            console.log("hello");
        },
    };
}
```

Is it displaying foo text and log hello?

**Reply by danie-ramdhani** [reply]
And downgrading livewire is not really helpful because i can't use new features.

**Reply by ekwoka** [reply]
I'm not saying that code does work or doesn't.

I'm saying Async Alpine is not part of Alpine, so I don't know what its issue would be.

The pure Alpine side seems to be all working fine.

**Reply by ekwoka** [reply]
oh, or this is related to the x-ignore issues mentioned elsewhere.

**Reply by ekwoka** [reply]
https://github.com/alpinejs/alpine/pull/4458

this is in 3.14.6...

I haven't looked that over...

**Reply by SimoTod** [reply]
nah, that won't fix the teleport issue. It was flagged by another user as well.
Teleport, for reasons I don't exactly know, has this line: 
https://github.com/alpinejs/alpine/blob/60cbc24ab6aff5b7699819565d64ca99f26c5c2b/packages/alpinejs/src/directives/x-teleport.js#L55

and, because of https://github.com/alpinejs/alpine/blob/463f2770ed5efb66b6487c9a33dfb7c8d116765d/packages/alpinejs/src/lifecycle.js#L87 which was added in a recent release, nothing can be initialised at a later stage inside a teleported element.

I have really little context about why the first line exists and why the second line was introduced but they can't really work well together.

**Reply by nikitaourazbaev** [reply]
Yes, I think I may be that other user, I reported this [here](https://github.com/alpinejs/alpine/discussions/4457#discussioncomment-11452670). We have an internal package that does similar things to Async Alpine: lazy-loading, etc.

I’m also wondering what the purpose of that `clone._x_ignore` is, and if it’s safe to remove it after it’s been added during the initial teleport?

It would be great to find a way to resolve the conflict between those two lines.

**Reply by SimoTod** [reply]
So x-teleport used to be the portal plugin. When it was a plugin, it did not have the ignore bit which was added when converted to a core directive here: https://github.com/alpinejs/alpine/pull/2431

Unfortunately, there's no context in that PR or in any of the commits and I don't think it's covered by test so it's not easy to predict what it would break if your remove that line.

 I suspect it's safe to remove since Alpine now marks elements that have already been processed so they don't get initialised twice so there's no point in ignoring the element after initTree but I can't be 100% sure.

**Reply by nikitaourazbaev** [reply]
Did some more testing today and I can actually run into this issue without lazy-loading or calling `initTree` from my own code in Alpine ^3.14.4.

Elements added to teleport targets after they’ve been cloned can’t initialize. 

Below is a link to a minimal repro using `x-if` that works on Alpine 3.14.3 but doesn’t work in later versions. From what I can tell it’s because `x-if` runs `initTree`, runs into the `_x_ignore` on the teleport target, and returns.

https://awesomealpine.com/play?share=RqrQ3-jaU_H7

@ekwoka I wonder if this is somehow related to the issue reported in #4461?

I know this example is contrived but this is actually something I’m running into in real-life code.

**Reply by joshhanley** [reply, marked as answer]
I had a look this with Caleb on our call this morning and we think the teleport `clone._x_ignore = true ` was something for Livewire. But we're not 100% on that. Caleb said to just try removing it and see if both test suites pass.

So I've submitted an Alpine PR #4469 with it removed and a Livewire PR https://github.com/livewire/livewire/pull/9072 with the Alpine fix and both test suites are passing.

**Reply by danie-ramdhani** [reply]
fixed in latest version, thanks @joshhanley !

**Reply by joshhanley** [reply]
@danie-ramdhani thanks for confirming 🙂 The fix is in v3.14.7 incase anyone else comes across this.

**Reply by nikitaourazbaev** [reply]
Fixed for me too, thanks @joshhanley!
